tensorizer/_linear_partition.py

```diff
@@ -51,108 +45,177 @@ def partition(
+        #   e) If earlier_weight is not too big, update best_weight,
+        #      jump back to the right, and go back to step 2
+        # This attempts to find the valid region for current_weight
+        # and earlier_weight as quickly as possible via a dynamic variant of
+        # binary search, and then linearly scans through the possibilities.
+
+        left = 0  # first impossible element due to current_weight
+        # Note: start = 0 is reserved for when preceding_parts = 0 anyway,
+        # which is always handled by the cache, so setting left = 0 is safe.
+        right = end - 1  # last possible element due to earlier_weight
+        start = right
+        while True:
+            current_weight = end_sum - prefix_sums[start]
+            if current_weight < best_weight:
+                earlier_weight = find_start(start, preceding_parts - 1)[1]
+                if earlier_weight < best_weight:
+                    best_weight = (
+                        current_weight
+                        if current_weight > earlier_weight
+                        else earlier_weight
                     )
-                    partition_starts[end - 1, partitions_remaining - 1] = start
-    del best_prefix_partition
-    slices = []
-    end = n - 1
-    for partitions_remaining in range(partitions - 1, 0, -1):
-        start = int(partition_starts[end - 1, partitions_remaining - 1])
-        # The algorithm calculates an inclusive endpoint,
-        # so end at end + 1 to include it, and start at start + 1
-        # to not overlap with the next slice
-        slices.append(slice(start + 1, end + 1))
-        end = start
-        if end == 0:
-            break
-    slices.append(slice(0, end + 1))
-    slices.reverse()
-
-    return slices
+                    best_start = start
+                    # If this was already the rightmost one, narrow the search
+                    right -= start == right
+                    if right <= left:
+                        break
+                    # Reset to the right end, in case the best was skipped
+                    start = right
+                else:
+                    # Nothing right of this matters
+                    # Try skipping forward a bit
+                    right = start - 1
+                    dist = right - left
+                    if dist <= 0:
+                        break
+                    else:
+                        start = right - (dist >> 1)
+            else:
+                # Overshot, nothing left of this matters
+                left = start
+                dist = right - left
+                if dist <= 0:
+                    break
+                elif dist == 1:
+                    start = right
+                else:
+                    start = left + (dist >> 1)
+
+        result = (best_start, best_weight)
+        memo[key] = result
+        return result
+
+    if partitions > 900:
```

Comment:
How'd we arrive at this number of partitions threshold?

Reply:
Python's recursion depth limit (`sys.getrecursionlimit()`) is 1000 by default. The recursive algorithm used here has one layer of recursion per partition, so anything near the default recursion depth limit triggers raising the limit up to some still-pretty-safe upper limit (~6500), and anything beyond that uses special logic to avoid any deeper recursive calls that the algorithm would use. For tensorizer, these are very unlikely to ever come up, since the number of partitions is at most the number of readers (which is also at most the number of tensors), and having 1000 concurrent readers in one process is most likely a bad idea, and should revert to using the greedy algorithm for partitioning because of the performance threshold anyway.

I rewrote an earlier version of this to do manual stack management with Python lists instead of function calls to avoid recursion to never have to deal with this problem (and its performance was pretty good), but the logic in this one is a little bit more involved, so the actually-recursive one seems nicer to keep here unless we need to change it.